### PR TITLE
Created Session Management.

### LINKED ISSUE #2356: Create an efficient session management system.
**Is your feature request related to a problem? Please describe.**
The current state of the Talawa app lacks a dedicated and efficient `session management mechanism`. Consequently, the `refreshToken` method and other resources related to user sessions are not being utilized optimally. This has resulted in several challenges:

`Token Refresh Overhead`:
The absence of a streamlined session management system leading to the unnecessary refresh of tokens before each API call. This redundancy poses an overhead on the client side, impacting the app's performance and responsiveness.

`Conflicts with Refresh Tokens`:
Inconsistent handling of token refreshes is leading to unintended behavior.

`App Crashes in Edge Cases`:
In certain edge use cases like `switching orgs repeatedly`, the absence of a session management strategy is causing `app crashes`. 

**Describe the solution you'd like**
Implement a proper and centralized session management mechanism in the application. So that the user sessions are handled effieciently.

**Describe alternatives you've considered**

**Approach to be followed (optional)**
A clear and concise description of approach to be followed.

**Additional context**

related issues:
- https://github.com/PalisadoesFoundation/talawa/issues/2342#issue-2092776141

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by Azad99-9**
**What kind of change does this PR introduce?**

This PR 

1. Introduces a new efficient Session Management System. 
2. Fixes the App crashes at multiple situations.

**Issue Number:**

Fixes #2356 

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

**If relevant, did you update the documentation?**

**Summary**

`Session Management Feature:`
Implemented a new `SessionManager` Class.

**Functionalities**:
1. Creates a User session as soon as the app launches.
2. Refreshes the refresh token at a regular interval, with a refresh duration of 10 min.

**advantages**:
1. Reduces the overhead of refreshing the refresh token prior to every api call.
2. Prevents app from crashing at multiple situations.

`Fixed App Crashes:`
Prior to this PR the Talawa app was prone to app crash in following situations:
1. Switching organisations.
2. Post login.
3. Post signup,
4. When app is put to rebuild in debug mode.
- All the above situations caused developers to reinstall the app every time they encounter an app crash.
     
All the above mentioned scenarios of app crashes were fixed in this PR.

**Does this PR introduce a breaking change?**

No

**Other information**

Created a new file `lib/service/session_manager.dart`

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by Azad99-9** [comment]
- Unauthorized file test failed because added a new package `fake_async.`
- Required for testing async methods that run asynchronously at `regular intervals`.

**Comment by Azad99-9** [comment]
@Cioppolo14 @noman2002 please assign reviewers for this PR.

**Comment by Cioppolo14** [comment]
@noman2002 @literalEval Can you review this PR?

**Comment by Dante291** [comment]
@Azad99-9 would this also solve the app crashing issue when user switch to other org, I am facing this frequently now

**Comment by Azad99-9** [comment]
Yes @Dante291 it covers all the situations of app crash that we are facing right now.

**Comment by Azad99-9** [comment]
From now on we need not to reinstall the app everytime we try to rebuild it.

**Comment by Dante291** [comment]
> From now on we need not to reinstall the app everytime we try to rebuild it.

wait but thats not likely 

> From now on we need not to reinstall the app everytime we try to rebuild it.

wait so like after this fix we need to reinstall the app every time we try to rebuild it? wouldn't that affect development environment as developer has to reinstall every time in order to work on it? or did I get this wrong

**Comment by Azad99-9** [comment]
Screenshot_20240116-001659

@Dante291 sorry you got me wrong, let me clear.

- The current app version we have on develop branch, crashes every time we try to rebuild it. As depicted in above image.
- This causes the devs to uninstall the app and reinstall it every time they face this crash, i guess you too have faced it? Making the developers work a bit hard.
- This PR has fixed aforementioned crash situation too.

File: lib/locator.dart [hunk 1]
```diff
@@ -75,6 +76,9 @@ final organizationService = locator<OrganizationService>();
 ///creating GetIt for ImageService.
 final imageService = locator<ImageService>();
 
+///creating GetIt for SessionManager.
```
Comment (literalEval) [inline lib/locator.dart:79]: Remove words like creating, returning etc.

File: lib/services/database_mutation_functions.dart [hunk 2]
```diff
@@ -39,6 +39,19 @@ class DataBaseMutationFunctions {
     _query = Queries();
   }
 
+  /// initializes [clientNonAuth] function.
```
Comment (literalEval) [inline lib/services/database_mutation_functions.dart:42]: Capitalize first letter.

File: lib/splash_screen.dart [hunk 3]
```diff
@@ -73,7 +73,7 @@ class _SplashScreenState extends State<SplashScreen> {
       setState(() => _initialUri = null);
     }
 
-    final bool userLoggedIn = await userConfig.userLoggedIn();
+    final bool userLoggedIn = userConfig.loggedIn;
```
Comment (literalEval) [inline lib/splash_screen.dart:76]: Cover this line in the tests.

**Review by literalEval — CHANGES_REQUESTED** [review]

**Comment by Dante291** [comment]
> Screenshot_20240131-225639
> 
> @Dante291 sorry you got me wrong, let me clear.
> 
> * The current app version we have on develop branch, crashes every time we try to rebuild it. As depicted in above image.
> * This causes the devs to uninstall the app and reinstall it every time they face this crash, i guess you too have faced it? Making the developers work a bit hard.
> * This PR has fixed aforementioned crash situation too.

Oh ok got it, yeah this fix is really needed, Thanks.

**Comment by Azad99-9** [comment]
apologies for work delay, I was out of station for few days.

File: pubspec.yaml [hunk 4]
```diff
@@ -69,11 +67,12 @@ dependencies:
...
   shimmer: ^3.0.0
   social_share: ^2.2.1
-  syncfusion_flutter_calendar: ^24.1.43
-  syncfusion_flutter_datepicker: 24.1.47
+  syncfusion_flutter_calendar: ^24.2.3
+  syncfusion_flutter_datepicker: 24.2.3
   timelines: ^0.1.0
   tutorial_coach_mark: ^1.2.11
   uni_links: ^0.5.1
+  uni_links_platform_interface: ^1.0.0
```
Comment (Azad99-9) [inline pubspec.yaml:75]: Added this line to mock platform interface in test environment.

**Comment by Azad99-9** [comment]
Screenshot from 2024-02-06 23-10-15

Even after covering the missing line the codecov is not being updated.

**Comment by Azad99-9** [comment]
> ## [Codecov](https://app.codecov.io/gh/PalisadoesFoundation/talawa/pull/2362?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation) Report
> Attention: `1 lines` in your changes are missing coverage. Please review.
> 
> > Comparison is base [(`d46880b`)](https://app.codecov.io/gh/PalisadoesFoundation/talawa/commit/d46880bad266539075c77b87d7d223d4d159b979?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation) 95.07% compared to head [(`6e26faf`)](https://app.codecov.io/gh/PalisadoesFoundation/talawa/pull/2362?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation) 95.07%.
> > Report is 5 commits behind head on develop.
> 
> > ❗ Current head [6e26faf](https://github.com/PalisadoesFoundation/talawa/commit/6e26faf61beb8ebb47aba25ca988b64d7a699eb4) differs from pull request most recent head [743b642](https://github.com/PalisadoesFoundation/talawa/commit/743b642dff52a8d4487a292264926ce673aaf0d3). Consider uploading reports for the commit [743b642](https://github.com/PalisadoesFoundation/talawa/commit/743b642dff52a8d4487a292264926ce673aaf0d3) to get more accurate results
> 
> [Files](https://app.codecov.io/gh/PalisadoesFoundation/talawa/pull/2362?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation)	Patch %	Lines
> [lib/splash_screen.dart](https://app.codecov.io/gh/PalisadoesFoundation/talawa/pull/2362?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation#diff-bGliL3NwbGFzaF9zY3JlZW4uZGFydA==)	0.00%	[1 Missing ⚠️ ](https://app.codecov.io/gh/PalisadoesFoundation/talawa/pull/2362?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation)
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/PalisadoesFoundation/talawa/pull/2362?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=PalisadoesFoundation).

Screenshot from 2024-02-06 23-10-15

Even after covering the missing line the codecov is not being updated.

**Comment by Azad99-9** [comment]
@literalEval please have a look at this PR.

**Review by noman2002 — APPROVED** [review]
Changes looks good to me. 
@Azad99-9 Good work.

**Review by palisadoes — APPROVED** [review]

**Comment by Azad99-9** [comment]
> Changes looks good to me. 
> @Azad99-9 Good work. 

Thank you for noticing my work @noman2002

**Pull request merged**